engines/director/score.cpp

```diff
@@ -1344,123 +1357,273 @@ void Score::playQueuedSound() {
+		_numChannels = _framesStream->readUint16();
+
+		if (_framesVersion > 13) {
+			_numChannelsDisplayed = _framesStream->readUint16();
 		} else {
-			if (framesVersion <= 7)	// Director5
+			if (_framesVersion <= 7)    // Director5
 				_numChannelsDisplayed = 48;
 			else
-				_numChannelsDisplayed = 120;	// D6
+				_numChannelsDisplayed = 120;    // D6
 
-			stream.readUint16(); // Skip
+			_framesStream->readUint16(); // Skip
 		}
 
-		size -= 2;
-
 		warning("STUB: Score::loadFrames. frame1Offset: %x numFrames: %x version: %x spriteRecordSize: %x numChannels: %x numChannelsDisplayed: %x",
-			frame1Offset, numFrames, framesVersion, spriteRecordSize, numChannels, _numChannelsDisplayed);
+			frame1Offset, _numFrames, _framesVersion, spriteRecordSize, _numChannels, _numChannelsDisplayed);
 		// Unknown, some bytes - constant (refer to contuinity).
 	}
 
-	uint16 channelSize;
-	uint16 channelOffset;
+	// TODO How to work with channelData that gets overridden
+	// partically by channels, hence we keep it and read the score from left to right
+	// TODO Merge it with shared cast
+	memset(_channelData, 0, kChannelDataSize);
 
-	Frame *initial = new Frame(this, _numChannelsDisplayed);
-	// Push a frame at frame#0 position.
 	// This makes all indexing simpler
-	_frames.push_back(initial);
+	_frameOffsets.push_back(0);
+
+	_currentTempo = 0;
+	_currentPaletteId = CastMemberID(0, 0);
+
+	// Prepare frameOffsets
+	_version = version;
+	_frameOffsets.push_back(_framesStream->pos());
+
+	// In case number of frames not known, precompute them!
+	if (_numFrames == 0) {
+		debugC(1, kDebugLoading, "Score::loadFrames(): Precomputing total number of frames!");
+		// Calculate number of frames beforehand.
+		int frameCount = 1;
+		while (loadFrame(frameCount)) {
+			debugC(1, kDebugLoading, "Score::loadFrames(): Skipped over frame %d!", frameCount);
+			frameCount++;
+		}
+		_numFrames = frameCount;
+		debugC(1, kDebugLoading, "Score::loadFrames(): Calculated, total number of frames %d!", _numFrames);
 
-	// This is a representation of the channelData. It gets overridden
-	// partically by channels, hence we keep it and read the score from left to right
-	//
-	// TODO Merge it with shared cast
-	byte channelData[kChannelDataSize];
-	memset(channelData, 0, kChannelDataSize);
+		memset(_channelData, 0, kChannelDataSize); // Reset channel data
+	}
 
-	uint8 currentTempo = 0;
-	CastMemberID currentPaletteId = CastMemberID(0, 0);
+	loadFrame(1);
 
-	while (size != 0 && !stream.eos()) {
-		uint16 frameSize = stream.readUint16();
-		debugC(3, kDebugLoading, "++++++++++ score frame %d (frameSize %d) size %d", _frames.size(), frameSize, size);
-		if (debugChannelSet(8, kDebugLoading)) {
-			stream.hexdump(frameSize);
+	// Read over frame offset array and print each item
+	for (uint i = 0; i < _frameOffsets.size(); i++) {
+		debugC(1, kDebugLoading, "Score::loadFrames(): Frame %d, offset %ld", i, _frameOffsets[i]);
```

Comment:
If you see any potential value in printing these offsets, it is fine, but definitely put it to debug level 7+. E.g. it will produce tons of noise.